applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -1187,81 +1247,65 @@ static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info
-	sys_snode_t *cur_node;
-	sys_snode_t *prev_node = NULL;
-
-	SYS_SLIST_FOR_EACH_NODE(&per->enqueued_out_reports, cur_node) {
-		struct enqueued_out_report *cur_report = CONTAINER_OF(cur_node,
-								      struct enqueued_out_report,
-								      node);
-
-		uint8_t cur_report_id = cur_report->data[0];
+	size_t orep_idx = get_output_report_idx(sent_report_id);
+	struct subscriber *sub = get_subscriber(per);
 
-		if (cur_report_id == sent_report_id) {
-			int send_err = send_hid_out_report(hogp, cur_report->data,
-							   cur_report->data_size);
+	if (!(per->enqueued_out_reports_bm & BIT(orep_idx))) {
+		return;
+	}
 
-			if (send_err) {
-				LOG_ERR("Cannot forward HID report (err: %d)", send_err);
-			}
+	__ASSERT_NO_MSG(sub->saved_out_reports_bm & BIT(orep_idx));
+	__ASSERT_NO_MSG(sent_report_id == sub->out_reports[orep_idx].report_id);
+	size_t size = get_output_report_full_size(sub->out_reports[orep_idx].report_id);
 
-			sys_slist_remove(&per->enqueued_out_reports, prev_node, cur_node);
-			k_free(cur_report->data);
-			k_free(cur_report);
-			break;
-		}
+	int send_err = send_hid_out_report(hogp, (uint8_t *)&sub->out_reports[orep_idx], size);
 
-		prev_node = cur_node;
+	if (send_err) {
+		LOG_ERR("Cannot forward HID report (err: %d)", send_err);
 	}
+	WRITE_BIT(per->enqueued_out_reports_bm, orep_idx, 0);
 }
 
-static void append_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+static void send_empty_hid_out_reports(struct hids_peripheral *per, struct subscriber *sub)
 {
-	struct enqueued_out_report *item = k_malloc(sizeof(*item));
+	struct bt_hogp *per_hogp = &per->hogp;
 
-	if (!item) {
-		LOG_ERR("Failed to allocate memory to forward HID output report");
-		return;
-	}
+	for (size_t orep_idx = 0; orep_idx < ARRAY_SIZE(sub->out_reports); orep_idx++) {
+		if (sub->saved_out_reports_bm & BIT(orep_idx)) {
+			uint8_t report_id = sub->out_reports[orep_idx].report_id;
+			size_t size = get_output_report_full_size(report_id);
+			uint8_t empty_data[size];
 
-	item->data_size = size;
-	item->data = k_malloc(size);
+			memset(empty_data, 0, sizeof(empty_data));
+			empty_data[0] = report_id;
+			int err = send_hid_out_report(per_hogp, empty_data, size);
 
-	if (!item->data) {
-		k_free(item);
-		LOG_ERR("Failed to allocate memory to forward HID output report");
-		return;
+			if (err) {
+				LOG_ERR("Failed to forward output report (err: %d)", err);
+			}
+		}
 	}
+}
 
-	memcpy(item->data, data, size);
-	sys_slist_append(list, &item->node);
-};
-
-static void enqueue_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+static void save_hid_out_report(struct subscriber *sub, const uint8_t *data, size_t size)
 {
 	uint8_t report_id = data[0];
-	bool updated = false;
+
+	__ASSERT_NO_MSG(size == get_output_report_full_size(report_id));
+	size_t orep_idx = get_output_report_idx(report_id);
 
 	/* If report is already present, update the data associated with report. */
```

Comment:
The report is always present in current implementation, please remove the comment :)